### DISCUSSION TITLE
Simpler device authorization end-user code verification for first-party clients

### DISCUSSION THREAD

**Discussion by alechenninger** [keycloak/keycloak#8829] [Ideas]
Currently, device authorization user code verification always displays the consent form, even for clients where consent is not required (e.g. first-party clients). The [spec](https://datatracker.ietf.org/doc/html/rfc8628) does not specifically require this. It is perhaps surprising to users (and/or admins) for clients which normally do not require consent.

In order to improve user experiences without meaningfully compromising security, I propose the following logic instead, based on whether or not the client is confidential, if consent is required, and if any scopes are requested (either because they are default scopes, or because they are optional scopes requested during authorization). Of course, for device authorization, it may be rare that a client is confidential, but I included this possibility for completeness sake.

| Is <br> confidential? | Consent <br> required? | Any scopes requested or defaulted? | Proposal |
|:--------------:|:------------------:|-------------------|----------|
| ✅  | ❌  | n/a | Device is authorized immediately on submission of a valid user code. User is displayed "device authorized" (success) landing page. |
| ❌  | ❌  | n/a | Client name must be confirmed after submission of user code. This is not the consent form–it does not include scopes–but instead a form just intended to confirm the client, in case the client is being impersonated from another application or device, per [rfc8628 section 5.6](https://datatracker.ietf.org/doc/html/rfc8628#section-5.6). This confirmation is not persistent and should be repeated each authorization for such a client. |
| ✅  | ✅ | ✅ | Show consent form if consent has not already been granted for this client. Otherwise, behave as if consent is not required (as additional consent _isn't_ required–consent has already been granted and persisted). |
| ✅  | ✅ | ❌ | Same as first case (see above for confidential client without consent required)  |
| ❌  | ✅ | ✅ | Show consent form every time (as currently implemented) per [rfc8252 section 8.6](https://datatracker.ietf.org/doc/html/rfc8252#section-8.6). |
| ❌  | ✅ | ❌ | Same as second case (see above for public client without consent required) |

Does anyone see any issues with this?

**Comment by stianst** [comment]
@mposolda @pedroigor FIY

**Comment by asumihasan** [comment]
Some in progress wireframes for the different scenarios:
![Device auth - Device login - KC](https://user-images.githubusercontent.com/67343292/143141456-ffd83143-bc67-49be-8c6e-af6e2f64cbe6.png)
![Device auth - Code entered - KC](https://user-images.githubusercontent.com/67343292/143141483-8d66c95e-f66f-4bc4-bcb1-f2d05edfe0ee.png)
![Device auth - Invalid code - KC](https://user-images.githubusercontent.com/67343292/143141492-61096d03-7db5-4fa2-abd3-652458867e38.png)
![Device auth - Confirm client - KC](https://user-images.githubusercontent.com/67343292/143141534-29bd3799-1ea4-4604-9a55-db4b0f30320b.png)
![Device auth - Grant access - KC](https://user-images.githubusercontent.com/67343292/143275452-77a4809b-902a-4a21-9b8d-c40aaaad58be.png)
![Device auth - Login successful - KC](https://user-images.githubusercontent.com/67343292/143141568-47c90e34-93b4-4c7f-bc69-bec3a9e561fb.png)
![Device auth - Login cancelled - KC](https://user-images.githubusercontent.com/67343292/143141575-5532bbe2-41ba-4ae1-8cfe-54aa529f311e.png)

**Comment by mposolda** [comment]
+1 for the proposal

@tnorimat FYI

**Comment by tnorimat** [comment]
@stianst @mposolda 
+1 for the proposal.

I've recognized that [requiring to confirm the client's identity by an end user every time](https://user-images.githubusercontent.com/67343292/143141534-29bd3799-1ea4-4604-9a55-db4b0f30320b.png) prevents keycloak from completing a device authorization grant automatically.

**Comment by alechenninger** [comment]
Because Keycloak does not show the consent form normally if there are no scopes requested, I clarified the cases where this happens. The novel scenario here is where there is a public client with consent required but no scopes requested (and no scopes are requested by default), we should still show the confirm client page as per the second scenario with a public client and no consent required.

**Reply by mposolda** [reply]
@alechenninger Thanks for this update!

**Comment by kherock** [comment]
Hi, we've recently become interested in this topic as we're exploring using Device Code grants for authenticating native apps without relying on deep links. Is this topic still active?

I think everything proposed so far makes sense, but we would like to still give our first-party clients that don't usually show consent screens special treatment to make the experience a bit smoother. Mainly, we don't want to show the consent-style format as we have that screen styled with the expectation that only 2nd & 3rd party integrations will see it.

I've suggested how clients that do vs don't require consent should differ below. I've also generalized the 3rd column to cover how the **Display client on screen** option currently affects how Keycloak renders consent screens.

Is confidential? | Consent required? | Any scopes or client to show on consent screen? | Proposal
-- | -- | -- | --
❌ | ❌ | n/a | Client name is confirmed after submission of user code, in case the client is being impersonated from another application or device, per [rfc8628 section 5.6](https://datatracker.ietf.org/doc/html/rfc8628#section-5.6). **This is not the consent form**–it does not include scopes–but instead a form just intended to confirm the client. This confirmation is not persistent and should be repeated each authorization for such a client.
❌ | ✅ | ❌ | Same as first case (request is treated as if client does not require consent)
❌ | ✅ | ✅ | Client name is confirmed via consent screen shown after submission of user code. This confirmation is not persistent and should be repeated each authorization for such a client per [rfc8252 section 8.6](https://datatracker.ietf.org/doc/html/rfc8252#section-8.6)
✅ | ❌ | n/a | Device is authorized immediately on submission of a valid user code. User is displayed "device authorized" (success) landing page.
✅ | ✅ | ❌ | Same as previous case (request is treated as if client does not require consent)
✅ | ✅ | ✅ | Show consent form if consent has not already been granted for this client. Otherwise, behave as if consent is not required (as additional consent isn't required–consent has already been granted and persisted).
